kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ServerTemplate.java

```diff
@@ -241,4 +236,48 @@ public int hashCode() {
         result = 31 * result + (capabilities != null ? capabilities.hashCode() : 0);
         return result;
     }
+
+    public ServerTemplate cloneServerTemplate() {
+        ServerTemplate serverTemplate = new ServerTemplate();
+        serverTemplate.setId(this.getId());
+        serverTemplate.setName(this.getName());
+        containersSpec.forEach(c -> serverTemplate.addContainerSpec(new ContainerSpec(c.getId(), c.getContainerName(), c.getServerTemplateKey(), c.getReleasedId(), c.getStatus(), cloneContainerConfigMap(c.getConfigs()))));
+
+        final Map<Capability, ServerConfig> cloneServerConfig = new HashMap<>();
+        configs.entrySet().stream().forEach(entry -> cloneServerConfig.put(entry.getKey(), entry.getValue()));
+        serverTemplate.setConfigs(cloneServerConfig);
+
+        serverInstances.forEach(serverInstance -> serverTemplate.addServerInstance(new ServerInstanceKey(serverInstance.getServerTemplateId(), serverInstance.getServerName(), serverInstance.getServerInstanceId(), serverInstance.getUrl())));
+
+        final List<String> cloneCapabilities = new ArrayList<>();
+        capabilities.forEach(capability -> cloneCapabilities.add(capability));
+        serverTemplate.setCapabilities(cloneCapabilities);
+
+        serverTemplate.setMode(mode);
+
+        return serverTemplate;
+    }
+
+    protected Map<Capability, ContainerConfig> cloneContainerConfigMap(final Map<Capability, ContainerConfig> configs) {
+        final Map<Capability, ContainerConfig> cloneConfigs = new HashMap<>();
+        configs.entrySet().stream().forEach(entry -> {
+            cloneConfigs.put(entry.getKey(), cloneContainerConfig(entry.getValue()));
```

Comment:
shallow copy may be good enough, but it need to change any object of ServerTemplate, My code just change one class.

Reply:
Not sure I understand that. Did you mean that in case a collection is added into them, we will need to reimplement the clone method? We will need to reimplement it anyway even with this code.